clients/web/src/collections/Collection.js

```diff
@@ -87,46 +129,138 @@ var Collection = Backbone.Collection.extend({
             return;
         }
 
+        if (this.filterFunc && !this.append) {
+            this.pageOffsetStack = this.pageOffsetStack || [];
+        }
+
         if (reset) {
+            if (this.filterFunc && !this.append) {
+                this.pageOffsetStack = [];
+            }
             this.offset = 0;
         } else {
             this.params = params || {};
         }
 
+        if (this.filterFunc && !this.append) {
+            this.pageOffsetStack.push(this.offset);
+        }
+
         var limit = this.pageLimit > 0 ? this.pageLimit + 1 : 0;
 
-        var xhr = restRequest({
-            path: this.altUrl || this.resourceName,
-            data: _.extend({
-                limit: limit,
-                offset: this.offset,
-                sort: this.sortField,
-                sortdir: this.sortDir
-            }, this.params)
-        }).done(_.bind(function (list) {
-            if (this.pageLimit > 0 && list.length > this.pageLimit) {
-                // This means we have more pages to display still. Pop off
-                // the extra that we fetched.
-                list.pop();
-                this._hasMorePages = true;
-            } else {
-                this._hasMorePages = false;
-            }
+        var finalList = []; /* will be built up in pieces */
 
-            this.offset += list.length;
+        function fetchListFragment() {
+            var xhr = restRequest({
+                path: this.altUrl || this.resourceName,
+                data: _.extend({
+                    limit: limit,
+                    offset: this.offset,
+                    sort: this.sortField,
+                    sortdir: this.sortDir
+                }, this.params)
+            });
 
-            if (list.length > 0 || reset) {
-                if (this.append && !reset) {
-                    this.add(list);
+            var result = $.when(xhr).then(_.bind(function (list) {
+                if (!_.isArray(list)) {
+                    list = [list];
+                }
```

Comment:
This is one change I don't have a good answer for, admittedly.  Before I added this check, tests where failing when trying to login/authenticate.  It seemed in those limited cases a collection fetch was returning a single scalar, which messed up the fetching algorithm.  After coercing scalar-returning fetches into length-1 arrays like I did above, everything started working again. so I left it in.

Maybe someone familiar with the user login tests could chime in and help me understand what was going on or possibly help me dig deeper to be absolutely sure if the array check is something we want to keep?